### ISSUE TITLE
FIlter alignments to display

### ISSUE THREAD

**Issue by arq5x** [igvteam/igv.js#1505]
Hi IGV team,

We are working more and more with very deep alignments (8000-50000X) and often want to focus solely on the small fraction of reads at a locus that differ from the reference. Is there a programmatic way to filter displayed alignments based on CIGAR or whether there is a difference at the center nucleotide in the display?  For example, if 10 of 10000 reads differ from the REF at position X, I'd want to ask IGV to not draw 9990 of the reads to improve performance and expedite inspection of the "interesting" 10.

**Comment by jrobinso** [comment]
Hi Aaron,  no there's not a way to do that now.   

Do you want to do this once during the load (filter alignments as they are loaded),  or dynamically and switch loci after the alignments are loaded.   If the former one can imagine supporting a filter function in the config of the track, that takes an alignment and returns a pass/fail flag.   If fail the alignment is discarded during loading.  Note this would affect coverage.

The semantics for the latter is more complex, and needs more definition.   How do you imagine programtically accessing the function?  That is do imagine calling a function, or setting something statically?  Some psuedo code might help.

The first option might help performance indirectly by freeing memory.   For both of these options you still have to load all the alignments.  Painting the alignments should be very quick,  only a portion of the alignments (vertically) are actually painted.

Normally the CIGAR won't tell if there's a mismatch to the reference, unless you are using special operators.

**Comment by jrobinso** [comment]
Yes I think that would be generally useful.     You screenshot implies a menu for user settings, we don't have this for any filters at the moment (for igv.js),  did you have in mind such a menu? 

Some random musing / observations. 

For visual purposes the filter could just control what is displayed, with everything in memory.  This allows rapid switching if there is a UI to support it,  but obviously does not yield any memory or performance advantages.  A filter which discards alignments would allow much larger regions to be shown,  as memory required would be minimal.   You still pay the penalty to load the alignments.

The MD string is not mandatory, btw,  as you know.  In the absense of the tag "M" does not imply no mismatches.  So we could either use the reference sequence if no MD tag,  or just disable this filter,  i.e. MD is required to use it.

There is the subtle question of when to filter,   the other filters you list filter before coverage counting.   I have a feeling this filter should be applied after coverage counting.

We might also want a variant of this filter for SVs,  something like only abnormal pairs, or only long reads with split alignments.

**Comment by arq5x** [comment]
`Yes I think that would be generally useful. You screenshot implies a menu for user settings, we don't have this for any filters at the moment (for igv.js), did you have in mind such a menu?`

I hadn't thought about how to do this for igv.js, but it certainly would be useful

`For visual purposes the filter could just control what is displayed, with everything in memory. This allows rapid switching if there is a UI to support it, but obviously does not yield any memory or performance advantages. A filter which discards alignments would allow much larger regions to be shown, as memory required would be minimal. You still pay the penalty to load the alignments.`

I see.  The main issue we have seen is the memory/performance associated with exploring alignments from very deep data. Perhaps this solution won't solve it in the way I had thought.

`There is the subtle question of when to filter, the other filters you list filter before coverage counting. I have a feeling this filter should be applied after coverage counting.`

Agreed

**Comment by jrobinso** [comment]
@arq5x You posted this to the igv.js issues forum,  are we talking about IGV desktop?   The discussion still applies, but implementation differs.

**Comment by arq5x** [comment]
My apologies - I was thinking about IGV, not igv.js, but such functionality for each would seemingly be useful.  Up to you, of course!
